src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -20,27 +28,74 @@
     private final PublicKeySignatureWriter signatureWriter = new PublicKeySignatureWriter();
 
     public boolean authenticate(String username, PublicKey key, ServerSession session) {
-        LOGGER.fine("Authentication attempted from "+username+" with "+key);
-        User u = User.get(username, false);
-        if (u==null) {
-            LOGGER.fine("No such user exists: "+username);
+        User user = this.retrieveOnlyKeyValidatedUser(username, key);
+
+        if (user == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
-        UserPropertyImpl sshKey = u.getProperty(UserPropertyImpl.class);
-        if (sshKey==null) {
-            LOGGER.fine("No SSH key registered for user: "+username);
+        Authentication auth = this.verifyUserUsingSecurityRealm(user);
+        if (auth == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
+        UserDetails userDetails = new SSHUserDetails(username, auth);
+        SecurityListener.fireAuthenticated(userDetails);
+        return true;
+    }
+
+    private @CheckForNull User retrieveOnlyKeyValidatedUser(String username, PublicKey key) {
+        LOGGER.log(Level.FINE, "Authentication attempted from {0} with {1}", new Object[]{ username, key });
+        User u = User.get(username, false, Collections.emptyMap());
```

Comment:
Use `getById`. We know we are not dealing with a `fullName`.

(No worse than the previous code, but so long as you are refactoring this you might as well use the more precise and efficient API.)

Reply:
Good to know, thank you 👍